[FEATURE] epoch() expression function

nirvn opened the pull request:
This PR adds an epoch() function to the expression engine which returns the interval in milliseconds between the unix epoch and a given date value.

For e.g., `epoch(to_date('2017-01-01'))` will return `1483203600000.`

It's a useful function to have, and can lead to exciting results when used alongside @nyalldawson 's color property assistant:
![untitled](https://cloud.githubusercontent.com/assets/1728657/23008544/e00834aa-f442-11e6-983f-0455f42b1892.png)
_Want to quickly shade polygons based on a date field? Sure can do!_

nirvn committed:
[FEATURE] epoch() expression function

This commit adds an epoch() function to the expression engine which
returns the interval in milliseconds between the unix epoch and a
given date value.

nirvn commented:
@m-kuhn , @NathanW2 , @wonder-sk , @nyalldawson , there appears to be a problem with the expression engine when dealing with int64 values. The following expression `epoch(now()) - 1` fails with the following error: Cannot convert '1487222866747' to int.

nyalldawson commented:
>  there appears to be a problem with the expression engine when dealing with int64 values. The following expression epoch(now()) - 1 fails with the following error: Cannot convert '1487222866747' to int.

This rings a bell about some PR which was closed which implemented long handling in expressions, but there was some deep issue with this which I can't recall... @m-kuhn I vaguely recall you creating the PR?

nirvn commented:
Here's the problem:

To do the subtraction, the expression engine checks whether it can transform epoch()'s QVariant( qint64 ) into a qint64 as part of the the getIntValue() function:

```
static int getIntValue( const QVariant& value, QgsExpression* parent )
{
  bool ok;
  qint64 x = value.toLongLong( &ok );
  if ( ok && x >= std::numeric_limits<int>::min() && x <= std::numeric_limits<int>::max() )
  {
    return x;
  }
  else
  {
    parent->setEvalErrorString( QObject::tr( "Cannot convert '%1' to int" ).arg( value.toString() ) );
    return 0;
  }
}
```

The problem is that std::numeric_limits<int>::max() returns `2147483647,` well below the actually qint64 epoch(now()) value of `1487229096468.`

Thoughts?

m-kuhn commented:
Apparently the check was introduced in 18451fdb to fix [issue #4622](https://hub.qgis.org/issues/4622).

Maybe the signature should be updated to return a qint64 instead and the boundary check removed?

nirvn commented:
@nyalldawson , @m-kuhn , ok to merge this and deal with the int vs qint64 issue separately?

Simply making the function return qint64 creates some issue with expression functions (such as char()) that need to pass int values to Qt functions.

nirvn commented:
@nyalldawson , @m-kuhn , here's a proposed solution: create a getInt64Value() function and use it when converting QVariant values to ints when dealing with operators. That way, expression functions requiring an actual int (such as char()) won't get impacted.

m-kuhn commented:
For consistency, can we use `qlonglong` and `getLongLongValue()`?

nirvn commented:
@m-kuhn , hmm -- can you look at the failing test and tell me what you think of it?

The expression  '-1'  returns a qint64 -1, which != int 1.

m-kuhn commented:
This one?

```
QDEBUG : TestQgsExpression::evaluation(unary minus int) got  qlonglong  instead of  int
FAIL!  : TestQgsExpression::evaluation(unary minus int) Compared values are not the same
   Loc: [/home/travis/build/qgis/QGIS/tests/src/core/testqgsexpression.cpp(1106)]
```

If it's just that `(int)-1` is now `qint64(-1)` that's expected and fine for me and the tests will need to be adjusted.

nirvn commented:
@m-kuhn ok, if you think the return value is fine I'll adjust the test.

nirvn commented:
An alternative approach would be to have a longlong check when dealing with operators (instead of always returning a longlong)

nirvn commented:
Hmm but that might complicate things. Ie can we do int + qint64? If not then we should keep current approach.

m-kuhn commented:
I think we should just deal with `qint64` (= `qlonglong`) by default and have a (less often used) `toNativeInt` method that can do checks and casts when required for some special calls like `char(x)`.

I don't see any good reason why we would want to deal with an int (possibly performance but I'm not convinced about that at all, 1. I'm not even sure there's a noticable impace, 2. I assume performance critical code shouldn't be using QVariants).
On the other hand, I see good reasons to always use `qlonglong` because of issues exactly like the once you are experiencing.

nyalldawson commented:
@wonder-sk do you remember earlier discussion about long support in expressions? I swear there was a massive thread or PR convo about it but can't find it now, and i have the sneaking suspicion that there was a tricky issue which blocked it in the past....

nirvn commented:
@m-kuhn , @nyalldawson , @wonder-sk , I've updated the commit to use qlonglong.

nirvn commented:
Ok, now that the transform curve is merged, I really, really want this pushed too 😉

nirvn commented:
OK, @wonder-sk and I discussed this off github and decided to settle on epoch() returning a double value (instead of qint64/qlonglong). Returning a double avoids the conversion-to-integer when doing subtractions / additions in the expression engine. 

This also removed the need to handle large int conversion for now, so I've reverted that specific commit.

I'll merge the PR when it turns green. Thanks to all for the feedback.

nirvn commented:
@nyalldawson, hmm, test passes on my machine but fails on travis (due to rounding during comparison). Any idea how to work around this travis issue?

nyalldawson commented:
try multiplying by 100 and then calling to_int

nyalldawson commented:
actually a lot more than 100 in this case :P

m-kuhn commented:
What's the problem with longlong support (apart from some dim memories of @nyalldawson ;) )?
It seems hacky to have epoch() return a double.

wonder-sk commented:
> It seems hacky to have epoch() return a double.

I suggested to use double - I do not see anything hacky about it :-) IMHO it avoids all the trouble with longlong support. It was the extra code for longlong that made me feel uncomfortable: some operations used longlong, but most operations just used simple int, making the implementation somehow inconsistent

m-kuhn commented:
If I understand that correctly, `epoch()` produces an integer. Returning a floating point number for something which is an integer by its nature seems hacky to me because it suggests higher precision than available, causes rounding issues, a small performance overhead, ... :-)

In QGIS 3 we could just switch the whole expression engine to longlong support. Just like python transparently handles long values.

```py
>>> type(14872228667478)
<type 'int'>
>>> type(1)
<type 'int'>
```

wonder-sk commented:
I just wanted to use double to avoid having to convert everything to longlong which I assume is much larger task (many functions to test for correct output). Double is also future proof if we ever decide to use microseconds etc. In the end I do not really care much whether it is longlong or double, anything that behaves consistently is fine...

m-kuhn commented:
Changing to double will just mask underlying issues which also surface when dealing with larger attribute integer values (supported by different providers).

I'd just return a longlong, accept that subtraction is broken for now and put it on the todo list for 3.0. How does that sound?

nirvn commented:
@m-kuhn , @wonder-sk , I can revert to a longlong, np. And indeed it'd be good to fix this lack of longlong support prior to 3.0.

nirvn commented:
Ok, the commit reverts to longlong -- for the record, this hack works until the underlying issue is fixed:

`to_real(epoch(now))) - 1`

m-kuhn commented:
@nirvn can you check with https://github.com/qgis/QGIS/pull/4176
